hail/src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -237,6 +237,10 @@ object Simplify {
     case ToStream(Let(name, value, ToArray(x)), _) if x.typ.isInstanceOf[TStream] =>
       Let(name, value, x)
 
+    //TODO: A better system would push all reshapes down to the MakeNDArray calls if there aren't any shape dependent nodes in between.
+    case NDArrayReshape(MakeNDArray(data, _, rowMajor), reshapedShape) => MakeNDArray(data, reshapedShape, rowMajor)
+    case NDArrayReshape(NDArrayMap(MakeNDArray(data, _, rowMajor), vName, body), newShape) => NDArrayMap(MakeNDArray(data, newShape, rowMajor), vName, body)
```

Comment:
You could make this closer to the goal in your TODO by making this rule just swap the reshape with the map, and let the previous rule do the rest.